python/lsst/pipe/base/graphBuilder.py

```diff
@@ -817,7 +819,79 @@ def makeGraph(self, pipeline, collections, run, userQuery):
+
+        userQuery = self._addInstrumentToQuery(pipeline, userQuery)
+
         with scaffolding.connectDataIds(self.registry, collections, userQuery) as commonDataIds:
             scaffolding.resolveDatasetRefs(self.registry, collections, run, commonDataIds,
                                            skipExisting=self.skipExisting)
         return scaffolding.makeQuantumGraph()
+
+    def _addInstrumentToQuery(self, pipeline, query):
+        """Add an instrument restriction to the query if the pipeline has an
+        instrument and the query does not already have an instrument. If the
+        query does have an instrument verify that it matches the instrument in
+        the pipeline.
+
+        Parameters
+        ----------
+        pipeline : `Pipeline`
+            Same as the pipeline in `makeGraph`
+        query : `str`
+            The initial query string.
+
+        Returns
+        -------
+        query : `str`
+            The query string with an instrument restriction checked or added
+            if needed.
+
+        Raises
+        ------
+        RuntimeError
+            If the query contains an instrument restriction and the pipeline
+            has an instrument and the two do not match.
+        """
+        instrument = pipeline.getInstrument()
+        if instrument is not None:
+            if isinstance(instrument, str):
+                instrument = getInstrument(pipeline._pipelineIR.instrument, self.registry)
+            queryInstrument = self._getInstrumentFromQuery(query)
+            if queryInstrument is None:
+                # There is not an instrument in the query, add it:
+                restriction = f"instrument = '{instrument.getName()}'"
+                _LOG.info(f"Adding restriction \"{restriction}\" to query.")
+                query = f"{restriction} AND ({query})"
+            elif queryInstrument != instrument.getName():
+                # Since there is an instrument in the query, it should match
+                # the instrument in the pipeline.
+                raise RuntimeError(f"The instrument named in the query (\"{queryInstrument}\") does not "
+                                   f"match the instrument named by the pipeline (\"{instrument.getName()}\")")
+        return query
+
+    @staticmethod
+    def _getInstrumentFromQuery(query):
+        """Search the provided query string for "instrument=" and if found
+        return the value that follows.
+
+        Search is case insensitive, and allows whitespace after "instrument"
+        before the equals sign.
+
+        Strips any quotes from the name of the instrument.
+
+        Parameters
+        ----------
+        query : `str`
+            The query string.
+
+        Returns
+        -------
+        instrumentName : `str` or `None`
+            The name of the instrument in the query string, or `None` if an
+            instrument is not named.
+        """
+        instrumentRegex = r"instrument *="
```

Comment:
As written this does not need to be a raw string. Saying that, in regexes I tend to prefer `\s*` rather than ` *` for whitespace matching.  Also though, why can't this regex also include the instrument in it as a group so that you can ask for group 1 directly, rather than taking the unmatched part and splitting on space after?

Something like:

```python
instrumentRegex =r"instrument\s*=\s*[\"'](.*?)[\"']"
```

works for me and the instrument is then `match.group(1)`.

Reply:
Even better:

    r"\Winstrument\s*=\s*[\"'](.*?)[\"']"

to avoid matching "NotAnInstrument = '1'".
I would actually suggest using expression parser that knows about query syntax to avoid surprises (ask me for more artificial examples if you want to know 🙂)

Reply:
And in our query language string literals use single quotes, like in SQL.

Reply:
I'd probably use `\binstrument` for word boundary...